### PR TITLE
CU-1u5ykup - add generated types and refactor, add eth sign to crowdloan tests, bonded finance tests implementation

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**
[I've followed the official documentation to generate all typescript definitions for dali](https://polkadot.js.org/docs/api/examples/promise/typegen/). Tests are now fully typed and IDE are able to infer endpoint/events types. I've improved the crowdloan rewards test, including association from both eth/relay accounts. This structure is probably not the final version as my knowledges of typescript/javascript stuff is near 0. Feel free to point anything that is wrongly done in this architecture.

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-1u5ykup Integration tests](https://app.clickup.com/t/1u5ykup)

**Review by Dom-Roth — APPROVED** [review]
Looks really good, and the generated types are going to make our lives a whole lot easier. Thanks for that. 
As a note for everyone else testing, the crowdloan associate tests are relying on #486.

File: integration-tests/runtime-tests/src/tests/tx/bondedFinance/txBondedFinanceTests.ts [hunk 1]
```diff
@@ -34,54 +22,24 @@ export class TxBondedFinanceTests {
...
+  private static txBondedFinanceOfferTest(wallet: IKeyringPair) {
     // ToDo: Find good parameter for reward.asset
     const requestParameters = {
       beneficiary: wallet.publicKey,
       asset: 1,
-      bondPrice: 1,
-      nbOfBonds: 10000000000000,
-      maturity: {Finite:{returnIn:50}},
-      reward: {asset: [170,141,183,460,469,231,731,687,303,715,884,105,728], amount: 10, maturity: 1}
+      bondPrice: 10000000000000, // pub const MIN_VESTED_TRANSFER: u32 = 1_000_000;
```
Comment (KaiserKarel) [inline integration-tests/runtime-tests/src/tests/tx/bondedFinance/txBondedFinanceTests.ts:30]: If we make `MIN_VESTED_TRANSFER` an associated constant, it becomes gettable over the API, and we can then not rely on magic constants in the tests

Comment (Dom-Roth) [reply integration-tests/runtime-tests/src/tests/tx/bondedFinance/txBondedFinanceTests.ts]: Updated!

File: integration-tests/runtime-tests/src/tests/tx/bondedFinance/txBondedFinanceTests.ts [hunk 2]
```diff
@@ -34,54 +22,24 @@ export class TxBondedFinanceTests {
...
       beneficiary: wallet.publicKey,
       asset: 1,
-      bondPrice: 1,
-      nbOfBonds: 10000000000000,
-      maturity: {Finite:{returnIn:50}},
-      reward: {asset: [170,141,183,460,469,231,731,687,303,715,884,105,728], amount: 10, maturity: 1}
+      bondPrice: api.consts.vesting.minVestedTransfer,
+      nbOfBonds: 1,
+      maturity: { Finite: { returnIn: 10 } },
+      reward: { asset: 1, amount: 100000000000000, maturity: 1 } // pub MinReward: Balance = 10 * CurrencyId::PICA.unit::<Balance>();
```
Comment (KaiserKarel) [inline integration-tests/runtime-tests/src/tests/tx/bondedFinance/txBondedFinanceTests.ts:33]: This one we could fetch from the API as well

**Comment by Dom-Roth** [comment]
I'll fix the unsigned commit on Monday, also fix a few things, as well as explain a bit more about the last commit.

**Comment by Dom-Roth** [comment]
As already announced, here are a few details about the last addition.

The main attraction of this commit is the implementation of extrinsic tests for the bondedFinance pallet.
Unfortunately are the bondedFinance.cancel tests failing so far, and I'm still searching for the source of this behavior.
I've modified **composable/frame/bonded-finance/src/lib.rs** and added a modifier to make MinReward accessible through a constant query.
```
#[pallet::constant] // <-- This
type MinReward: Get<BalanceOf<Self>>;
```

The query test file (**queryCrowdloanRewardsTets.ts**) was updated to reflect the new test style.

The query test files **queryTokenTests.ts** and **querySystemAccountTests.ts** have been removed.
The **queryTokenTests.ts** file was partially a stub and stopped working with the new type implementation.
The **querySystemAccountTests.ts** file also stopped working with the new type implementation.
Both of those files will be re-added, when the time has come.

I've also added a configuration file for mocha (**.mocharc.json**). To get rid of passed configurations with start parameters.
And also some general cleanup.

**Here are the current results of the tests when run.**
```
  query.crowdloanRewards.account Tests
  ✓ query.crowdloanRewards.claimedRewards Tests
  queryCrowdloanRewardsTotalContributorsTest
  ✓ query.crowdloanRewards.totalContributors Tests
  ✓ query.crowdloanRewards.totalRewards Tests
  
  CrowdloanRewards Tests
  ✓ Can populate the list of contributors (23311ms)
  ✓ Can initialize the crowdloan (23919ms)
  1) Can associate a picasso account
  
  tx.bondedFinance Success Tests
  ✓ Can create a new offer (24007ms)
  ✓ Can create a second new offer (23969ms)
  ✓ Can bond to newly created offer (23950ms)
  
  tx.bondedFinance Failure Tests
  ✓ Should not be able to create offer (bondPrice < MIN_VESTED_TRANSFER) (24037ms)
  ✓ Should not be able to create offer (reward.amount < MinReward) (23974ms)
  
  tx.bondedFinance.cancel Success Tests
  2) Can cancel offer created in first bondedFinance.offer test by creator
  3) Can sudo (diff. account) cancel offer created in second bondedFinance.offer
  
  tx.bondedFinance.cancel Failure Tests
  ✓ Should not be able to cancel offer that doesn't exist (15854ms)
```

The first failed test: crowdloanRewards.associate fails, since it depends on #486 as mentioned above, and I wasn't running that server instance.
2,3 are bondedFinance.cancel tests of which I'm not sure yet why they're failing, but I'm working on it. (Maybe @hussein-aitlahcen  has an idea?)

For questions or improvements, just comment here or contact me anytime. :)

**Comment by Dom-Roth** [comment]
### Last commit summary:

Added a change to the bondedFinance pallet. In the cancel function, the second transfer call now has keep-alive=false. Cause it made the function fail, and i suppose this is due to the staking wallet, getting deleted when the offer gets cancelled.

The test functions of **txBondedFinanceTests.ts** were moved into their own folders, to reduce the file length. And a few more tests were added.

Updated the **README.md** file, to not show an infinite timeout as an example, and added run information for the type generator.

**Comment by Dom-Roth** [comment]
### Last Test Results

```
query.crowdloanRewards.account Tests
    ✓ query.crowdloanRewards.claimedRewards Tests
queryCrowdloanRewardsTotalContributorsTest
    ✓ query.crowdloanRewards.totalContributors Tests
    ✓ query.crowdloanRewards.totalRewards Tests

  CrowdloanRewards Tests
    ✓ Can populate the list of contributors (12518ms)
    ✓ Can initialize the crowdloan (24008ms)
    ✓ Can associate a picasso account (23999ms)

  tx.bondedFinance.offer & .bond Success Tests
    ✓ Can create a new offer (23995ms)
    ✓ Can create a second new offer (23978ms)
    ✓ Can bond to newly created offer (23976ms)

  tx.bondedFinance.offer Failure Tests
    ✓ Should not be able to create offer (bondPrice < MIN_VESTED_TRANSFER) (24044ms)
    ✓ Should not be able to create offer (reward.amount < MinReward) (23995ms)
    ✓ Should not be able to create offer (reward.asset does not exist) (23982ms)

  tx.bondedFinance.cancel Failure Tests
    ✓ Should not be able to cancel offer that doesn't exist (23988ms)

  tx.bondedFinance.cancel Success Tests
    ✓ Can cancel offer created in first bondedFinance.offer test by creator (24013ms)
    ✓ Can sudo (diff. account) cancel offer created in second bondedFinance.offer (24003ms)
```

**Comment by hussein-aitlahcen** [comment]
> ### Last Test Results
> ```
> query.crowdloanRewards.account Tests
>     ✓ query.crowdloanRewards.claimedRewards Tests
> queryCrowdloanRewardsTotalContributorsTest
>     ✓ query.crowdloanRewards.totalContributors Tests
>     ✓ query.crowdloanRewards.totalRewards Tests
> 
>   CrowdloanRewards Tests
>     ✓ Can populate the list of contributors (12518ms)
>     ✓ Can initialize the crowdloan (24008ms)
>     ✓ Can associate a picasso account (23999ms)
> 
>   tx.bondedFinance.offer & .bond Success Tests
>     ✓ Can create a new offer (23995ms)
>     ✓ Can create a second new offer (23978ms)
>     ✓ Can bond to newly created offer (23976ms)
> 
>   tx.bondedFinance.offer Failure Tests
>     ✓ Should not be able to create offer (bondPrice < MIN_VESTED_TRANSFER) (24044ms)
>     ✓ Should not be able to create offer (reward.amount < MinReward) (23995ms)
>     ✓ Should not be able to create offer (reward.asset does not exist) (23982ms)
> 
>   tx.bondedFinance.cancel Failure Tests
>     ✓ Should not be able to cancel offer that doesn't exist (23988ms)
> 
>   tx.bondedFinance.cancel Success Tests
>     ✓ Can cancel offer created in first bondedFinance.offer test by creator (24013ms)
>     ✓ Can sudo (diff. account) cancel offer created in second bondedFinance.offer (24003ms)
> ```

Nice!

File: frame/crowdloan-rewards/Cargo.toml [hunk 3]
```diff
@@ -74,13 +69,10 @@ std = [
 
 runtime-benchmarks = [
     "hex-literal",
-    "once_cell",
     "libsecp256k1",
     "balances",
```
Comment (KaiserKarel) [inline frame/crowdloan-rewards/Cargo.toml:73]: should this be balances/runtime-benchmarks?

Comment (Dom-Roth) [reply frame/crowdloan-rewards/Cargo.toml:73]: This was from a commit I pulled from main. -> #485

**Review by KaiserKarel — DISMISSED** [review]

**Review by KaiserKarel — APPROVED** [review]

**Review by Dom-Roth — APPROVED** [review]

**Pull request merged**